



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.1.0">
    
    
      
        <title>Tutorial - CASPR</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.11e41852.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.22915126.css">
      
      
        
        
        <meta name="theme-color" content="#4caf50">
      
    
    
      <script src="../assets/javascripts/modernizr.20ef595d.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="green" data-md-color-accent="indigo">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#tutorial" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="CASPR" class="md-header-nav__button md-logo">
          
            <img src="../images/logo.svg" width="24" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                CASPR
              </span>
              <span class="md-header-nav__topic">
                Tutorial
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/judithbergada/CASPR/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      judithbergada/CASPR
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="CASPR" class="md-nav__button md-logo">
      
        <img src="../images/logo.svg" width="48" height="48">
      
    </a>
    CASPR
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/judithbergada/CASPR/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      judithbergada/CASPR
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../usage/" title="Usage" class="md-nav__link">
      Usage
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Tutorial
      </label>
    
    <a href="./" title="Tutorial" class="md-nav__link md-nav__link--active">
      Tutorial
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#get-started-activate-conda-environment" title="Get started: Activate Conda Environment" class="md-nav__link">
    Get started: Activate Conda Environment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#user-case-1-analysis-of-a-crispr-screen-using-pgrnas" title="User Case 1: Analysis of a CRISPR Screen Using pgRNAs" class="md-nav__link">
    User Case 1: Analysis of a CRISPR Screen Using pgRNAs
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#results" title="Results" class="md-nav__link">
    Results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#user-case-2-analysis-of-a-crispr-screen-using-sgrnas" title="User Case 2: Analysis of a CRISPR Screen Using sgRNAs" class="md-nav__link">
    User Case 2: Analysis of a CRISPR Screen Using sgRNAs
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#results_1" title="Results" class="md-nav__link">
    Results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#get-started-activate-conda-environment" title="Get started: Activate Conda Environment" class="md-nav__link">
    Get started: Activate Conda Environment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#user-case-1-analysis-of-a-crispr-screen-using-pgrnas" title="User Case 1: Analysis of a CRISPR Screen Using pgRNAs" class="md-nav__link">
    User Case 1: Analysis of a CRISPR Screen Using pgRNAs
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#results" title="Results" class="md-nav__link">
    Results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#user-case-2-analysis-of-a-crispr-screen-using-sgrnas" title="User Case 2: Analysis of a CRISPR Screen Using sgRNAs" class="md-nav__link">
    User Case 2: Analysis of a CRISPR Screen Using sgRNAs
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#results_1" title="Results" class="md-nav__link">
    Results
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/judithbergada/CASPR/tree/master/docs/pages/tutorial.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="tutorial">Tutorial</h1>
<p>Running CASPR is extremely easy and convenient to analyze CRIPR-Cas9 screens
using pgRNAs. The test data folder contains two brief user cases to go through
all steps. Simply copy the scripts provided in each user case to visualize the
demos.</p>
<h2 id="get-started-activate-conda-environment">Get started: Activate Conda Environment</h2>
<p>If it is your first time with CASPR, please follow the
<a href="../installation/">Installation Guide</a>.</p>
<p>After finishing with the installation, you should activate the Conda
Environment. Please, make sure you do it every time you work with CASPR:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Activate for Linux</span>
<span class="nb">source</span> activate caspr
<span class="c1"># Activate for macOS</span>
conda activate caspr
</pre></div>

<div class="admonition info">
<p class="admonition-title">Users of IBU Cluster</p>
<p>Users of IBU cluster (Interfaculty Bioinformatics Unit, University of Bern)
will also download CASPR following the
<a href="../installation/">Installation Guide</a>.</p>
<p>However, you do not need to poceed with the installation of a Conda
Environment. Instead, you can load the required software packages,
which are all installed in the cluster.</p>
<p>Please, make sure you load them every time you work with CASPR:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Import software: only for users of IBY cluster</span>
module add UHTS/Analysis/fastx_toolkit/0.0.13.2
module add UHTS/Aligner/STAR/2.6.0c
module add UHTS/Analysis/samtools/1.4
module add R/3.5.1
module add UHTS/Quality_control/mageck-vispr/0.5.4
</pre></div>

</div>
<h2 id="user-case-1-analysis-of-a-crispr-screen-using-pgrnas">User Case 1: Analysis of a CRISPR Screen Using pgRNAs</h2>
<p>This user case shows how to analyse a CRISPR screen with the following features:</p>
<ul>
<li>Library of <strong>pgRNAs</strong></li>
<li>Two time points: week zero and week four</li>
<li>Only one replicate per time point</li>
<li>Start analysis from the raw sequencing data</li>
</ul>
<p>To compute the analysis, it is essential to have:</p>
<ul>
<li>Paired reads</li>
<li>Text file with a library of pgRNAs</li>
<li>Text file with the experimental design (control and treated samples)</li>
</ul>
<p>Additionally, you may provide a text file with positive, negative and neutral
controls.</p>
<p>Let's go to the user case test data folder:</p>
<div class="codehilite"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/CASPR/testdata/pgrna
</pre></div>

<p>Let's check that the data is available in the expected formats:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Check that the data has been sucessfully downoaded</span>
ls -al .

<span class="c1"># Display the content of the files to see the format</span>
cat expdesign.txt
head library.txt
head controlfile.txt
zcat &lt; week0.1_1.fastq.gz <span class="p">|</span> head
</pre></div>

<p>Now you are ready to run the example. If you are not sure about the commands,
you can copy them immediately from here:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Create output directory</span>
mkdir results

<span class="c1"># Run the tool</span>
CASPR <span class="se">\</span>
  -f <span class="s2">&quot;week0.1_1.fastq.gz, week4.1_1.fastq.gz&quot;</span> <span class="se">\</span>
  -r <span class="s2">&quot;week0.1_2.fastq.gz, week4.1_2.fastq.gz&quot;</span> <span class="se">\</span>
  --library library.txt <span class="se">\</span>
  -y <span class="m">0</span>.25 -k <span class="se">\</span>
  --output-dir ./results <span class="se">\</span>
  --controls controlfile.txt <span class="se">\</span>
  --exper-design expdesign.txt
</pre></div>

<p>Note that, if you copy the commands from above, you are changing two of the
default parameters of CASPR:</p>
<ol>
<li>the FDR threshold <code>-y</code> will be set to 0.25, and</li>
<li>the intermediate files will be kept.</li>
</ol>
<p>Please, make sure you keep the intermediate files to follow the second part of
the user case.</p>
<h3 id="results">Results</h3>
<p>After a few minutes, the outputs should appear in your computer. At this point,
you can open them and see if they are as expected. The expected outputs are
found in:</p>
<div class="codehilite"><pre><span></span>ls results/outputs
</pre></div>

<p>These are the outpus generated:</p>
<ul>
<li><a href="../assets/pgrna/Alignment_statistics.pdf">Alignment Statistics</a></li>
</ul>
<p><img src="../assets/pgrna/Alignment_statistics.png" alt="Alignment Statistics" width="40%"></p>
<ul>
<li><a href="../assets/pgrna/Comp_MAGeCK_PBNPA.pdf">Comparison of MAGeCK and PBNPA</a></li>
</ul>
<p><img src="../assets/pgrna/Comp_MAGeCK_PBNPA.png" alt="Comparison MAGeCK-PBNPA" width="40%"></p>
<ul>
<li><a href="../assets/pgrna/Log_counts.pdf">Log Counts</a></li>
</ul>
<p><img src="../assets/pgrna/Log_counts-0.png" alt="Log Counts" width="50%"><img src="../assets/pgrna/Log_counts-1.png" alt="Log Counts" width="50%"></p>
<ul>
<li><a href="../assets/pgrna/Selected_genes_MAGeCK.pdf">Selected Genes by MAGeCK</a></li>
</ul>
<p><img src="../assets/pgrna/Selected_genes_MAGeCK-0.png" alt="Selected Genes by MAGeCK" width="33.3%"><img src="../assets/pgrna/Selected_genes_MAGeCK-1.png" alt="Selected Genes by MAGeCK" width="33.3%"><img src="../assets/pgrna/Selected_genes_MAGeCK-2.png" alt="Selected Genes by MAGeCK" width="33.3%"></p>
<ul>
<li><a href="../assets/pgrna/Selected_genes_PBNPA.pdf">Selected Genes by PBNPA</a></li>
</ul>
<p><img src="../assets/pgrna/Selected_genes_PBNPA-0.png" alt="Selected Genes by PBNPA" width="50%"><img src="../assets/pgrna/Selected_genes_PBNPA-1.png" alt="Selected Genes by PBNPA" width="50%"></p>
<ul>
<li><a href="../assets/pgrna/Trimming_statistics.pdf">Trimming Statistics</a></li>
</ul>
<p><img src="../assets/pgrna/Trimming_statistics-2.png" alt="Trimming Statistics" width="33.3%"><img src="../assets/pgrna/Trimming_statistics-3.png" alt="Trimming Statistics" width="33.3%"><img src="../assets/pgrna/Trimming_statistics-4.png" alt="Trimming Statistics" width="33.3%"></p>
<p>Moreover, you can benefit from VISPR to get interactive results on you data.
Please try the following:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Run VIPR server</span>
vispr server ./results/config* --port <span class="m">5000</span>
<span class="c1"># Ctrl-C to exit the server</span>
</pre></div>

<div class="admonition info">
<p class="admonition-title">Users of IBU Cluster</p>
<p>VISPR installation in the IBU cluster might not be updated. Therefore,
users of this cluster will need to run an additional command to see
the outputs:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Take VISPR from jpijuan Conda Environment</span>
<span class="nb">alias</span> <span class="nv">vispr</span><span class="o">=</span><span class="s2">&quot;/home/jpijuan/miniconda3/envs/vispr/bin/vispr&quot;</span>

<span class="c1"># Run VIPR server</span>
vispr server ./results/config* --port <span class="m">5000</span>
<span class="c1"># Ctrl-C to exit the server</span>
</pre></div>

</div>
<p>VISPR is a web-based interactive server that creates a webpage with a nice visualization of the results.
If you are working from a cluster, you will also need to run this command locally on your computer:</p>
<div class="codehilite"><pre><span></span>ssh -f <span class="o">{</span>user<span class="o">}</span>@binfservms01.unibe.ch -L <span class="m">5000</span>:localhost:5000 -N
</pre></div>

<p>Now, check the <a href="http://127.0.0.1:5000">webpage</a> on a browser and enjoy visualizing
the data.</p>
<p>If you are still curious about other options of CASPR, let's try to get more
information on the unmapped reads. This can be done esily just by adding the tag
<code>-i</code> to the previous command. Furthermore, if you kept the intermediate files
with the <code>-k</code> argument, neither the previous nor the next steps will be
necessary anymore (only mapping).</p>
<p>You can find a solution to quickly finish the user case here:</p>
<div class="codehilite"><pre><span></span>CASPR <span class="se">\</span>
  -f <span class="s2">&quot;week0.1_1.fastq.gz, week4.1_1.fastq.gz&quot;</span> <span class="se">\</span>
  -r <span class="s2">&quot;week0.1_2.fastq.gz, week4.1_2.fastq.gz&quot;</span> <span class="se">\</span>
  --library library.txt <span class="se">\</span>
  -i --start map --pause map -y <span class="m">0</span>.25 -k <span class="se">\</span>
  --output-dir ./results <span class="se">\</span>
  --controls controlfile.txt <span class="se">\</span>
  --exper-design expdesign.txt
</pre></div>

<p>Check that everything worked out. Your outputs should look similar to:</p>
<ul>
<li><a href="../assets/pgrna/Alignment_statistics_2.pdf">Alignment Statistics</a></li>
</ul>
<p><img src="../assets/pgrna/Alignment_statistics_2-0.png" alt="Alignment Statistics" width="40%"><img src="../assets/pgrna/Alignment_statistics_2-1.png" alt="Alignment Statistics" width="40%"></p>
<p>If you arrived here, you finished the first user case successfully!</p>
<h2 id="user-case-2-analysis-of-a-crispr-screen-using-sgrnas">User Case 2: Analysis of a CRISPR Screen Using sgRNAs</h2>
<p>This user case shows how to analyse a CRISPR screen with the following features:</p>
<ul>
<li>Library of <strong>sgRNAs</strong></li>
<li>Two time points: week zero and week four</li>
<li>Two replicates per time point</li>
<li>Start analysis from the raw sequencing data</li>
</ul>
<p>To compute the analysis, it is essential to have:</p>
<ul>
<li>Single-read sequencing</li>
<li>Text file with a library of sgRNAs</li>
<li>Text file with the experimental design (control and treated samples)</li>
</ul>
<p>Moreover, you may provide a text file with positive, negative and neutral
controls, if interested.</p>
<p>In this user case, we want to analyse the two replicates separately in two
different tests, and we want to perform a third test using the replicates
together.</p>
<p>Let's go to the user case test data folder:</p>
<div class="codehilite"><pre><span></span><span class="nb">cd</span> <span class="nv">$HOME</span>/CASPR/testdata/sgrna
</pre></div>

<p>Before starting the tests, check that the data is available in the expected
formats. Pay special attention on the experimental design file, which allows to
obtain the results as required:</p>
<div class="codehilite"><pre><span></span>cat expdesign.txt
head library.txt
head controlfile.txt
zcat &lt; week0.1.fastq.gz <span class="p">|</span> head
</pre></div>

<p>At this point, you are probably ready to try the analysis yourself. To get nicer
visualization of the outputs, it is recommended to use an FDR threshold of 0.25.</p>
<p>If you need further help, you can also use these commands:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Create output directory</span>
mkdir results

<span class="c1"># Run the tool</span>
CASPR <span class="se">\</span>
  -f <span class="s2">&quot;week0.1.fastq.gz, week0.2.fastq.gz, week4.1.fastq.gz, week4.2.fastq.gz&quot;</span> <span class="se">\</span>
  --library library.txt <span class="se">\</span>
  -y <span class="m">0</span>.25 <span class="se">\</span>
  --output-dir ./results <span class="se">\</span>
  --controls controlfile.txt <span class="se">\</span>
  --exper-design expdesign.txt
</pre></div>

<h3 id="results_1">Results</h3>
<p>Quickly, the outputs should appear in your computer. You can open them and see
if they look as expected. The expected outputs are found in:</p>
<div class="codehilite"><pre><span></span>ls results/outputs
</pre></div>

<p>These are the outpus generated:</p>
<ul>
<li><a href="../assets/sgrna/Alignment_statistics.pdf">Alignment Statistics</a></li>
</ul>
<p><img src="../assets/sgrna/Alignment_statistics.png" alt="Alignment Statistics" width="40%"></p>
<ul>
<li><a href="../assets/sgrna/Comp_MAGeCK_PBNPA.pdf">Comparison of MAGeCK and PBNPA</a></li>
</ul>
<p><img src="../assets/sgrna/Comp_MAGeCK_PBNPA-2.png" alt="Comparison MAGeCK-PBNPA" width="40%"></p>
<ul>
<li><a href="../assets/sgrna/Log_counts.pdf">Log Counts</a></li>
</ul>
<p><img src="../assets/sgrna/Log_counts-4.png" alt="Log Counts" width="50%"><img src="../assets/sgrna/Log_counts-5.png" alt="Log Counts" width="50%"></p>
<ul>
<li><a href="../assets/sgrna/Selected_genes_MAGeCK.pdf">Selected Genes by MAGeCK</a></li>
</ul>
<p><img src="../assets/sgrna/Selected_genes_MAGeCK-6.png" alt="Selected Genes by MAGeCK" width="33.3%"><img src="../assets/sgrna/Selected_genes_MAGeCK-7.png" alt="Selected Genes by MAGeCK" width="33.3%"><img src="../assets/sgrna/Selected_genes_MAGeCK-8.png" alt="Selected Genes by MAGeCK" width="33.3%"></p>
<ul>
<li><a href="../assets/sgrna/Selected_genes_PBNPA.pdf">Selected Genes by PBNPA</a></li>
</ul>
<p><img src="../assets/sgrna/Selected_genes_PBNPA-4.png" alt="Selected Genes by PBNPA" width="50%"><img src="../assets/sgrna/Selected_genes_PBNPA-5.png" alt="Selected Genes by PBNPA" width="50%"></p>
<ul>
<li><a href="../assets/sgrna/Trimming_statistics.pdf">Trimming Statistics</a></li>
</ul>
<p><img src="../assets/sgrna/Trimming_statistics-1.png" alt="Trimming Statistics" width="33.3%"><img src="../assets/sgrna/Trimming_statistics-3.png" alt="Trimming Statistics" width="33.3%"><img src="../assets/sgrna/Trimming_statistics-4.png" alt="Trimming Statistics" width="33.3%"></p>
<p>Finally, let's perfom the test again using an FDR threshold of 0.1. For that,
you do not need to execute all the previous steps, only the assessment of gene
significance.</p>
<p>It can be done like this:</p>
<div class="codehilite"><pre><span></span>CASPR <span class="se">\</span>
  --start <span class="nb">test</span> -y <span class="m">0</span>.1 -k <span class="se">\</span>
  --output-dir ./results <span class="se">\</span>
  --controls controlfile.txt <span class="se">\</span>
  --exper-design expdesign.txt
</pre></div>

<p>The results should be exactly the same as above, but fewer hits are displayed in
the plots. This option ensures that you can perform analyses of CRISPR screens
not only starting rom the raw sequencing data, but also beginning from a table
of read counts.</p>
<p>If you kept the intermediate files in this last step, you can visualize again
the results with VISPR:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Run VIPR server</span>
vispr server ./results/config* --port <span class="m">6000</span>
<span class="c1"># Ctrl-C to exit the server</span>
</pre></div>

<div class="admonition info">
<p class="admonition-title">Users of IBU Cluster</p>
<p>VISPR installation in the IBU cluster might not be updated. Therefore,
users of this cluster will need to run an additional command to see
the outputs:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Take VISPR from jpijuan Conda Environment</span>
<span class="nb">alias</span> <span class="nv">vispr</span><span class="o">=</span><span class="s2">&quot;/home/jpijuan/miniconda3/envs/vispr/bin/vispr&quot;</span>

<span class="c1"># Run VIPR server</span>
vispr server ./results/config* --port <span class="m">6000</span>
<span class="c1"># Ctrl-C to exit the server</span>
</pre></div>

</div>
<p>Also keep in mind that, if you work from a cluster, you will need to copy the following
line locally on your computer in order to see the results:</p>
<div class="codehilite"><pre><span></span>ssh -f <span class="o">{</span>user<span class="o">}</span>@binfservms01.unibe.ch -L <span class="m">6000</span>:localhost:6000 -N
</pre></div>

<p>The data should be already available at this <a href="http://127.0.0.1:6000">webpage</a>.</p>
<p>Now that you arrived here, you are totally ready to use CASPR with your CRISPR data.
Enjoy!</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../usage/" title="Usage" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Usage
              </span>
            </div>
          </a>
        
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2018 Judith Bergadà Pijuan
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
  <div class="md-footer-social">
    <link rel="stylesheet" href="../assets/fonts/font-awesome.css">
    
      <a href="https://github.com/judithbergada/CASPR" class="md-footer-social__link fa fa-github-alt"></a>
    
  </div>

      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.9e1f3b71.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
    
    
      
        <script>!function(e,a,t,n,o,c,i){e.GoogleAnalyticsObject=o,e.ga=e.ga||function(){(e.ga.q=e.ga.q||[]).push(arguments)},e.ga.l=1*new Date,c=a.createElement(t),i=a.getElementsByTagName(t)[0],c.async=1,c.src="https://www.google-analytics.com/analytics.js",i.parentNode.insertBefore(c,i)}(window,document,"script",0,"ga"),ga("create","UA-139052696-1","auto"),ga("set","anonymizeIp",!0),ga("send","pageview");var links=document.getElementsByTagName("a");if(Array.prototype.map.call(links,function(a){a.host!=document.location.host&&a.addEventListener("click",function(){var e=a.getAttribute("data-md-action")||"follow";ga("send","event","outbound",e,a.href)})}),document.forms.search){var query=document.forms.search.query;query.addEventListener("blur",function(){if(this.value){var e=document.location.pathname;ga("send","pageview",e+"?q="+this.value)}})}</script>
      
    
  </body>
</html>